lib/util/fast_jason_encoder.ex

```diff
@@ -0,0 +1,68 @@
+# Copyright(c) 2015-2019 ACCESS CO., LTD. All rights reserved.
+
+defmodule Antikythera.FastJasonEncoder do
+  @moduledoc """
+  This module converts structures to JSON at low load by avoiding protocol calls. Internally jason is used.
+  """
+  alias Antikythera.Time
+
+  defmodule Wrapper do
+    defstruct item: nil
+  end
+
+  defimpl Jason.Encoder, for: Wrapper do
+    def encode(%Wrapper{item: item}, opts) do
+      Antikythera.FastJasonEncoder.encode(item, opts)
+    end
+  end
+
+  def encode(value) do
+    Jason.encode(%Wrapper{item: value})
+  end
+
+  def encode(value, opts) when is_list(value) do
+    case value do
+      []   -> "[]"
+      list -> [?[ | encode_list(list, opts, false)]
+    end
+  end
+  def encode(value, opts) when is_map(value) do
+    case Map.to_list(value) do
+      []   -> "{}"
+      list -> [?{ | encode_map(list, opts, false)]
+    end
+  end
+  def encode({Time, _ymd, _hms, _ms} = time, opts) do
+    if Time.valid?(time) do
+      [?", Time.to_iso_timestamp(time), ?"]
+    else
+      Jason.Encode.value(time, opts)
+    end
+  end
+  def encode(value, opts) do
+    Jason.Encode.value(value, opts)
+  end
```

Comment:
(optional)
If you have more than 3 function branches, you can simplify it (at least for me) by using `Croma`'s multi-clause function definition.
```ex
defun encode(value :: any, opts :: Jason.Encode.opts) :: iodata do
  ([]       , _   )                       -> "[]"
  (list     , opts) when is_list(list)    -> [?[ | encode_list(list, opts, false)]
  (empty_map, _   ) when empty_map == %{} -> "{}"
  (map      , opts) when is_map(map)      -> [?{ | encode_map(Map.to_list(map), opts, false)]
  (value    , opts)                       ->
    if Time.valid?(value) do
      [?", Time.to_iso_timestamp(value), ?"]
    else
      Jason.Encode.value(value, opts)
    end
end
```

This is also true for `encode_map/3` and `encode_list/3`.

Reply:
refactored. 71d7dca5ed1cfaa9f7855e7b4794ec21c4ae05fa